examples/ctc.dx

```diff
@@ -0,0 +1,134 @@
+'## Connectionist Temporal Classification
+By Alex Graves et alia.
+
+'[Link to paper](https://www.cs.toronto.edu/~graves/icml_2006.pdf).
+
+' This dynamic programming algorithm computes the probability of a particular sequences of labels
+(without pauses included) given another sequence of label probabilities
+(with pauses included), marginalizing over all possible combination of
+pause lengths.
+It's used for training speech-to-text models on unaligned training data.
+
+'Most implementations of CTC compute the marginal logprob
+and its gradients combining a forward and a backward pass.
+However, as mentioned in the original paper, we should only
+need the last step of the forward pass to compute the marginal
+probability, and Dex's autodiff should produce the backward
+pass automatically.  That makes this code much shorter than
+most implementations.
+
+
+def interleave (blank:v) (labels: m=>v) : (m & (Fin 2))=>v =
+  -- Turns "text" into "t e x t " by first pairing each letter with a blank,
+  -- then flattening the pairs back into a single-index table.
+  pairs = for i. [labels.i, blank]
+  for (i, j). pairs.i.j
+
+def prepend (first: v) (seq: m=>v) : (Unit|m)=>v =
+  -- Concatenates a single element to the beginning of a sequence.
+  for idx. case idx
+    Left () -> first
+    Right i -> seq.i
+
+def prepend_and_interleave (blank:v) (seq: m=>v) : 
+  ((Unit | (m & (Fin 2)))=>v) =
+  -- Turns "text" into " t e x t".
+  -- The output of this function has a slightly complicated output type, which
+  -- has size 1 + 2 * (size m).
+  interleaved = interleave blank seq
+  prepend blank interleaved
+
+def clipidx (n:Type) -> (i:Int) : n =
+  -- Returns element at 0 if less than zero.
+  -- Ideally we could have an alternative
+  -- to Fin that just clips the index at its bounds.
+  fromOrdinal n (select (i < 0) 0 i)
+
+def logaddexp (x:Real) (y:Real) : Real =
+  m = max x y
+  m + ( log ( (exp (x - m) + exp (y - m))))
+
+def ctc (time : Type) ?-> (vocab : Type) ?-> (position : Type) ?->
+        (dict: Eq vocab) ?=> (dict2: Eq position) ?=> (dict3: Eq time) ?=> (blank: vocab)
+        (logits: time=>vocab=>Real) (labels: position=>vocab) : Real =
+  -- Computes log p(labels | logits), marginalizing over possible alignments.
+  -- Todo: remove unnecessary implicit type annotations once
+  -- Dex starts putting implicit types in scope.
+
```

Comment:
Yes, I agree that putting the implicit implicit arguments in scope will reduce a lot of clutter. Making the dictionaries implicit will be harder but perhaps we can at least make it syntactically lighter. Even now, you can use `_` instead of names like `dict`.

Reply:
Ok.  Also, if I can get rid of some of the ```select i == j``` then I think I need only 1 Eq dictionary.